File: Divide_&_Conqeur/169_Majority_Event/solution169.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int res = 0, count = 0;

        for (int n : nums) {
            if (count == 0) {
                res = n;
            }
            count += (n == res) ? 1 : -1;
        }
        
        return res;
    }
};

// Runtime1: 15ms, beats 77.13%
// Memory1: 26.92MB, beats 79.70%

// Runtime2: 13ms, beats 83.67%
// Memory2: 27.04MB, beats 49.73%

// Runtime3: 15ms, beats 77.13%
// Memory3: 26.92MB, beats 79.70%

File: Divide_&_Conqeur/169_Majority_Event/solution169.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        res, count = 0, 0

        for n in nums:
            if count == 0:
                res = n
            count += (1 if n == res else -1)
            
        return res

# Runtime1: 169ms, beats 64.87%
# Memory1: 18.01MB, beats 24.65%

# Runtime2: 158ms, beats 96.30%
# Memory2: 18.09MB, beats 24.65%

# Runtime3: 170ms, beats 60.33%
# Memory3: 18.07MB, beats 24.65%